### PR TITLE
don't hydrate on unknown routes

### PR DESCRIPTION
**Pull request by andrew-nowak**
Attempting to hydrate causes a blank page to load when viewed through web.archive.org, viewing 'cached results' from Google, internal page monitoring tools and probably more

**Overall change:** Don't attempt to hydrate if the window location is unknown

**Alternative solutions considered:** Choosing whether to hydrate based on `window.location.hostname` - this has the benefit that it would allow us to add a page at `/search` without breaking the google cache view page

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -9,9 +10,11 @@ import { template, templateStyles } from './app/lib/joinUsTemplate';
 const data = window.SIMORGH_DATA || {};
 const root = document.getElementById('root');
 
-Loadable.preloadReady().then(() => {
-  hydrate(<ClientApp data={data} routes={routes} />, root);
-});
+if (matchRoutes(routes, window.location.pathname).length > 0) {
```
Comment (jamesbhobbs) [inline src/client.js]: Can we add an inline comment to explain why we're doing this, I can easily imagine someone coming along and removing it

**Review by dr3 — COMMENTED** [review]
We currently only serve error pages for routes we accept, but this would mean, if we ever serve error pages for all routes, like https://github.com/bbc/simorgh/issues/1544 will do, This means we wont hydrate those error pages, which I think we would want to do. 

Its gunna be a long time until its possible for a user to hit a simorgh page that doesnt have an explict route, are we ok not hydrating them when that happens? would a simple `onClient()` wrap here work too?

**Comment by jamesdonoh** [comment]
Good catch @aacn500. I quite like the idea of validating `window.location.hostname` _as well_, as if only to put an extra hurdle in the way of would-be phishers/fake news disseminators squatting on `www.not-the-real.bbbc.com` (though you could argue that's easily worked around).

That reminds me, do we have a way of globally disabling CSR yet?

**Review by dr3 — DISMISSED** [review]
https://github.com/bbc/simorgh/pull/2612#pullrequestreview-267032907

**Comment by andrew-nowak** [comment]
@dr3 ah I wasn't aware that we were going to start doing that. yeah that would become an issue, and I think we would have to change the check in this pr to inspecting `window.location.hostname`.

That was going to be my preferred solution anyway, except that I wanted to whitelist `*.bbc.co{.uk,m}` (to allow `www.`, `www.test.`, `routing.`, etc. etc.) until @hindsc52 pointed out that the monitoring team's tool is suffering the same issue and is hosted on an internal bbc.co.uk domain.

I think then we would need to whitelist domains for hydration _without wildcarding_. I doubt that I'm aware of all the domains we would need to support, would anyone be able to help me in figuring those out?

**Comment by dr3** [comment]
Idm about the domain restrictions, I dont think we need to do a route match though

edit: didnt fully understand the issue, will have a think about this

**Comment by andrew-nowak** [comment]
Closing for now; will reopen if any future ways of solving this are discovered

**Comment by andrew-nowak** [comment]
it's back!

A new alternate solution: when SSRing, put the path into SIMORGH_DATA. when hydrating, check that that path and window.location match. if not, we're not running on 'real' simorgh, but a cached/archived page and hydration will probably fail, so don't

File: src/client.js [hunk 2]
```diff
@@ -9,9 +9,16 @@ import { template, templateStyles } from './app/lib/joinUsTemplate';
...
+// Only hydrate the client if we're on a known route
+// When on an unknown route, the SSR will be discarded and the user will only
+// see a blank screen. This should never happen on BBC domains, but web page
+// archival/caching sites such as Google webcache and web.archive.org, among
+// others.
+if (window.SIMORGH_DATA.path === window.location.pathname) {
+  Loadable.preloadReady().then(() => {
+    hydrate(<ClientApp data={data} routes={routes} />, root);
+  });
+}
```
Comment (dr3) [inline src/client.js:29]: Can we add a logger info call to this on an `else` please? If this breaks "Is there a console log" will be an easy debug imo

**Review by chris-hinds — APPROVED** [review]
👍 pending the addition of the logger output suggested by @dr3

File: src/client.js [hunk 3]
```diff
@@ -9,9 +9,22 @@ import { template, templateStyles } from './app/lib/joinUsTemplate';
...
+// When on an unknown route, the SSR would be discarded and the user would only
+// see a blank screen. Avoid this by only hydrating when the embedded page data
+// and window location agree what the path is. Otherwise, fallback to the SSR.
+if (window.SIMORGH_DATA.path === window.location.pathname) {
+  Loadable.preloadReady().then(() => {
+    hydrate(<ClientApp data={data} routes={routes} />, root);
+  });
+} else {
+  // eslint-disable-next-line no-console
+  console.log(`
```
Comment (dr3) [inline src/client.js]: Soz to be a pain, can you use `logger.node` please? it has a client swap in that will log for us

Comment (andrew-nowak) [reply src/client.js]: :+1:

File: src/client.js [hunk 4]
```diff
@@ -19,7 +22,7 @@ if (window.SIMORGH_DATA.path === window.location.pathname) {
   });
 } else {
   // eslint-disable-next-line no-console
```
Comment (j-pendlebury) [inline src/client.js]: is this still needed? logger is not a console.log 😄

Comment (andrew-nowak) [reply src/client.js]:
```suggestion
```

**Review by j-pendlebury — APPROVED** [review]
LGTM 👍

**Comment by jamesbhobbs** [comment]
The testing on this should be extremely specific. What MUST be checked for is the hydration on every browser we support. If the page does hydrate I doubt anything else could go wrong, but as you'll be doing lot's of cross-browser tests anyhow a quick check of the content of the page would be sensible.

**Comment by PriyaKR** [comment]
Looks good to me across browsers and devices.Needs verifying on feature phones.

**Comment by PriyaKR** [comment]
Verified on feature phone looks good.Good to be merged.

**Pull request merged**